balancer/pickfirstleaf/test/pickfirstleaf_test.go

```diff
@@ -0,0 +1,731 @@
+}
+
+func Test(t *testing.T) {
+	grpctest.RunSubTests(t, s{})
+}
+
+// setupPickFirstLeaf performs steps required for pick_first tests. It starts a
+// bunch of backends exporting the TestService, creates a ClientConn to them
+// with service config specifying the use of the pick_first LB policy.
+func setupPickFirstLeaf(t *testing.T, backendCount int, opts ...grpc.DialOption) (*grpc.ClientConn, *manual.Resolver, *backendManager) {
+	t.Helper()
+	r := manual.NewBuilderWithScheme("whatever")
+	backends := make([]*stubserver.StubServer, backendCount)
+	addrs := make([]resolver.Address, backendCount)
+
+	for i := 0; i < backendCount; i++ {
+		backend := &stubserver.StubServer{
+			EmptyCallF: func(_ context.Context, _ *testpb.Empty) (*testpb.Empty, error) {
+				return &testpb.Empty{}, nil
+			},
+		}
+		if err := backend.StartServer(); err != nil {
+			t.Fatalf("Failed to start backend: %v", err)
+		}
+		t.Logf("Started TestService backend at: %q", backend.Address)
+		t.Cleanup(func() { backend.Stop() })
+
+		backends[i] = backend
+		addrs[i] = resolver.Address{Addr: backend.Address}
+	}
+
+	dopts := []grpc.DialOption{
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithResolvers(r),
+		grpc.WithDefaultServiceConfig(stateStoringServiceConfig),
+	}
+	dopts = append(dopts, opts...)
+	cc, err := grpc.NewClient(r.Scheme()+":///test.server", dopts...)
+	if err != nil {
+		t.Fatalf("grpc.NewClient() failed: %v", err)
+	}
+	t.Cleanup(func() { cc.Close() })
+
+	// At this point, the resolver has not returned any addresses to the channel.
+	// This RPC must block until the context expires.
+	sCtx, sCancel := context.WithTimeout(context.Background(), defaultTestShortTimeout)
+	defer sCancel()
+	client := testgrpc.NewTestServiceClient(cc)
+	if _, err := client.EmptyCall(sCtx, &testpb.Empty{}); status.Code(err) != codes.DeadlineExceeded {
+		t.Fatalf("EmptyCall() = %s, want %s", status.Code(err), codes.DeadlineExceeded)
+	}
+	return cc, r, &backendManager{backends}
+}
+
+// TestPickFirstLeaf_SimpleResolverUpdate tests the behaviour of the pick first
+// policy when when given an list of addresses. The following steps are carried
+// out in order:
+//  1. A list of addresses are given through the resolver. Only one
+//     of the servers is running.
+//  2. RPCs are sent to verify they reach the running server.
+//
+// The state transitions of the ClientConn and all the subconns created are
+// verified.
+func (s) TestPickFirstLeaf_SimpleResolverUpdate(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	balChan := make(chan *stateStoringBalancer, 1)
+	balancer.Register(&stateStoringBalancerBuilder{balancer: balChan})
+	tests := []struct {
+		name                     string
+		backendCount             int
+		backendIndexes           []int
+		runningBackendIndex      int
+		wantSCStates             []scStateExpectation
+		wantConnStateTransitions []connectivity.State
+	}{
+		{
```

Comment:
I still feel this table makes these tests harder to read than they have to be. I tried splitting them into separate tests and this is what I get:

This is the first test in the table. I also simplified this even further by not stopping the second backend since it is immaterial.

```golang
func (s) TestPickFirstLeaf_SimpleResolverUpdate1(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
	defer cancel()
	balChan := make(chan *stateStoringBalancer, 1)
	balancer.Register(&stateStoringBalancerBuilder{balancer: balChan})

	cc, r, bm := setupPickFirstLeaf(t, 2)
	addrs := stubBackendsToResolverAddrs(bm.backends)
	stateSubscriber := &ccStateSubscriber{transitions: []connectivity.State{}}
	internal.SubscribeToConnectivityStateChanges.(func(cc *grpc.ClientConn, s grpcsync.Subscriber) func())(cc, stateSubscriber)

	r.UpdateState(resolver.State{Addresses: addrs})
	bal := <-balChan
	testutils.AwaitState(ctx, t, cc, connectivity.Ready)

	if err := pickfirst.CheckRPCsToBackend(ctx, cc, addrs[0]); err != nil {
		t.Fatal(err)
	}

	wantSCStates := []scStateExpectation{{ServerIdx: 0, State: connectivity.Ready}}
	if diff := cmp.Diff(scStateExpectationToSCState(wantSCStates, addrs), bal.subConns()); diff != "" {
		t.Errorf("subconn states mismatch (-want +got):\n%s", diff)
	}

	wantConnStateTransitions := []connectivity.State{
		connectivity.Connecting,
		connectivity.Ready,
	}
	if diff := cmp.Diff(wantConnStateTransitions, stateSubscriber.transitions); diff != "" {
		t.Errorf("ClientConn states mismatch (-want +got):\n%s", diff)
	}
}
```

The other tests in the table also end up as small as this one.
```golang
func (s) TestPickFirstLeaf_SimpleResolverUpdate2(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
	defer cancel()
	balChan := make(chan *stateStoringBalancer, 1)
	balancer.Register(&stateStoringBalancerBuilder{balancer: balChan})

	cc, r, bm := setupPickFirstLeaf(t, 2)
	addrs := stubBackendsToResolverAddrs(bm.backends)
	bm.backends[0].Stop()

	stateSubscriber := &ccStateSubscriber{transitions: []connectivity.State{}}
	internal.SubscribeToConnectivityStateChanges.(func(cc *grpc.ClientConn, s grpcsync.Subscriber) func())(cc, stateSubscriber)

	r.UpdateState(resolver.State{Addresses: addrs})
	bal := <-balChan
	testutils.AwaitState(ctx, t, cc, connectivity.Ready)

	if err := pickfirst.CheckRPCsToBackend(ctx, cc, addrs[1]); err != nil {
		t.Fatal(err)
	}

	wantSCStates := []scStateExpectation{
		{ServerIdx: 0, State: connectivity.Shutdown},
		{ServerIdx: 1, State: connectivity.Ready},
	}
	if diff := cmp.Diff(scStateExpectationToSCState(wantSCStates, addrs), bal.subConns()); diff != "" {
		t.Errorf("subconn states mismatch (-want +got):\n%s", diff)
	}

	wantConnStateTransitions := []connectivity.State{
		connectivity.Connecting,
		connectivity.Ready,
	}
	if diff := cmp.Diff(wantConnStateTransitions, stateSubscriber.transitions); diff != "" {
		t.Errorf("ClientConn states mismatch (-want +got):\n%s", diff)
	}
}
```

and
```
func (s) TestPickFirstLeaf_SimpleResolverUpdate3(t *testing.T) {
	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
	defer cancel()
	balChan := make(chan *stateStoringBalancer, 1)
	balancer.Register(&stateStoringBalancerBuilder{balancer: balChan})

	cc, r, bm := setupPickFirstLeaf(t, 2)
	addrs := stubBackendsToResolverAddrs(bm.backends)
	bm.backends[0].Stop()

	stateSubscriber := &ccStateSubscriber{transitions: []connectivity.State{}}
	internal.SubscribeToConnectivityStateChanges.(func(cc *grpc.ClientConn, s grpcsync.Subscriber) func())(cc, stateSubscriber)

	addrs = append([]resolver.Address{addrs[0]}, addrs...)
	r.UpdateState(resolver.State{Addresses: addrs})
	bal := <-balChan
	testutils.AwaitState(ctx, t, cc, connectivity.Ready)

	if err := pickfirst.CheckRPCsToBackend(ctx, cc, addrs[2]); err != nil {
		t.Fatal(err)
	}

	wantSCStates := []scStateExpectation{
		{ServerIdx: 0, State: connectivity.Shutdown},
		{ServerIdx: 2, State: connectivity.Ready},
	}
	if diff := cmp.Diff(scStateExpectationToSCState(wantSCStates, addrs), bal.subConns()); diff != "" {
		t.Errorf("subconn states mismatch (-want +got):\n%s", diff)
	}

	wantConnStateTransitions := []connectivity.State{
		connectivity.Connecting,
		connectivity.Ready,
	}
	if diff := cmp.Diff(wantConnStateTransitions, stateSubscriber.transitions); diff != "" {
		t.Errorf("ClientConn states mismatch (-want +got):\n%s", diff)
	}
}
```

After writing the test for the third entry in the table though, I would also essentially get rid of the `scStateExpectationToSCState` function and instead do a literal struct initialization for the expected subchannel states since that would be more readable.